### DISCUSSION TITLE
Preserve ICC profile while converting from TIFF to JXL

### DISCUSSION THREAD

**Discussion by Qhilm** [libvips/libvips#4211] [Q&A] [answered]
Hello,

I am using this command to losslessly (from the point of view of image quality) convert a TIFF to JXL:

`vips jxlsave image.tif image.jxl --lossless`

I notice that the ICC info is missing in the resulting file. Any idea how to preserve it? exiftool doesn't seem to be capable of copying ICC info from TIFF to JXL.

libvips 8.15.5, installed using brew.

**Comment by jcupitt** [comment]
Hello @Qhilm,

ICC profile support for JXL is a new feature in libvips 8.16. It's in rc2 right now, it should be out in a few days.

https://github.com/libvips/libvips/releases/tag/v8.16.0-rc2

**Reply by Qhilm** [reply]
fantastic news.

My use case involves the CLI tool `vips`: will it give control over ICC profile (copy, transform...) or is it currently only planned as API in `libvips`? E.g.: when converting a tiff to jxl, also convert ICC profile to Display P3.

Forgive my ignorance, I don't know anything about the feature parity level between `vips` command line and `libvips`.

**Reply by jcupitt** [reply]
It should work fine from the CLI -- open a bug report if you have problems.

The libvips CLI is generated automatically from the libvips API, so there ought to be complete parity.

**Reply by jcupitt** [reply]
> E.g.: when converting a tiff to jxl, also convert ICC profile to Display P3.

It copies the ICC profile over, so a P3 profile on the TIFF will become a P3 profile in JXL. You can do colour space transforms with `icc_transform` during the copy if you need to.

**Reply by Qhilm** [reply]
Thanks. Sorry for the very basic question, but `icc_transform` requires "Filename to load output profile from" according to the help (I'm using v8.15.5 at the moment)

Can't I simply specify "Display P3" as output ICC profile?

I feel I'm missing something basic, apologies in advance for this and many thanks for taking the time to help me out.

**Reply by jcupitt** [reply]
You'd do:

```
vips icc_transform x.tif y.jxl p3
```

The `p3` will make it use the P3 ICC profile that ships with libvips. 

This will map the RGB pixels in the source TIFF to P3 colourspace using the input profile embedded in the TIFF (assuming sRGB if there is no embedded profile) and the P3 profile as the output. It won't change the way the image looks, just the way it is encoded.

**Reply by Qhilm** [reply]
thanks, but this command definitely changes the way the picture looks, it performs a lossy compression in my tests. I want to keep it lossless.
My current approach would be:
```
vips icc_transform image.tif image-p3.tif p3
vips jxlsave image-p3.tif image-p3.jxl --lossless
```

If I can summarise this into a single command (in libvips 8.16), that would of course be even better.

**Reply by jcupitt** [reply, marked as answer]
Yes, `jxlsave` in 8.15 won't attach the p3 profile, so it'll come out looking desaturated and dark. In 8.16 it should look identical.

You can write it as one command in 8.15 as well:

```
vips icc_transform x.tif y.jxl[lossless] p3
```

**Reply by Qhilm** [reply]
sorry I wasn't clear: I meant that my 50MB tif file became  a 1MB jxl file after `vips icc_transform x.tif y.jxl p3`

Hence I believe that `icc_transform` is compressing the image, not simply replacing the icc profile.

**Reply by jcupitt** [reply]
Yes, you need the `[lossless]` tag on the jxl filename.

**Reply by Qhilm** [reply]
ah yes, sorry, I read the first sentence and thought you misunderstood me and didn't notice the `[lossless]` tag.

ok, this works for me, thanks a lot.

I'm not clear on why there is a need for the `jxlsave` command additionally, if I can do all of this with `icc_transfrom` already?

**Reply by jcupitt** [reply]
The `vips` command really executes three libvips operations: one to load the image (it picks a loader for you based on the image signature), one for the operation, and one for the saver (it uses the filename suffix to pick a saver).

So this command:

```
vips icc_transform x.tif y.jxl[lossless] p3
```

Is really executing this:

```
tiffload x.tif | icc_transform p3 | jxlsave y.jxl --lossless
```

The `--flag` arguments to the `vips` command are given to the middle operation (`icc_transform`), and you give arguments to the loader and saver with `[square,brackets]`.

It's very handy for scripting because you can write things like this:

```bash
#!/bin/bash

vips icc_transform $1 /tmp/t1.v p3
vips crop /tmp/t1.v $2 0 0 100 100
```

And run that script with:

```
./convert_and_crop.sh ~/pics/x.tif[page=2] ~/pics/y.jxl[lossless]
```

ie. convert page 2 of that tiff to a lossless JXL. You can give arguments to any loader or saver with no extra programming.

There's a chapter in the docs about this:

https://www.libvips.org/API/current/using-cli.html

**Reply by Qhilm** [reply]
thanks for taking the time to write this detailed explanation!

I think icc_transform is then already sufficient for me and I do not need jxlsave 8.16.

**Reply by jcupitt** [reply]
You'll need 8.16 to get an icc profile in the output JXL image. Without it, your p3 images will look pretty bad.

**Comment by Qhilm** [comment]
I can't see any issue with converting from sRGB to P3 it seems, see attached files. Both images look identical on my screen (whose screen does use a P3 profile, it's a MacBook). I mean, not a big deal, but I fear it's a symptom that I'm missing something very basic here...

`vips icc_transform a.jpg "b-icc_convert-p3.jxl[lossless,effort=9]" p3` (libvips 8.15.5)

Secondly, it seems libvips is not using the `lossless_jpeg` cjxl option, as the resulting jxl file is significantly larger than the original jpg, instead of being smaller. But I'll open a separate discussion for this, in order not to mix up topics.

![a](https://github.com/user-attachments/assets/d7752e3f-922d-48bc-bac1-955b4b722640)
![b-cjxl jxl](https://github.com/user-attachments/assets/d1924a34-96ad-41a0-b54e-7685f121c2a8)
